views/clickhouse.go

```diff
@@ -125,17 +107,45 @@ func buildRealtimeViewsEventsQuery(spec QuerySpec) (string, []interface{}, error
 	return sql, args, nil
 }
 
-func filterMostRecent(res []RealtimeViewershipRow) []RealtimeViewershipRow {
-	if len(res) > 0 {
-		var filtered []RealtimeViewershipRow
-		mostRecentTimestamp := res[0].Timestamp
-		for _, r := range res {
-			if mostRecentTimestamp == r.Timestamp {
-				filtered = append(filtered, r)
-			} else {
-				return filtered
-			}
+// currentEventsQuery uses an unoptimized raw playback logs to favor latency over query speed.
+func currentEventsQuery(spec QuerySpec) squirrel.SelectBuilder {
+	return squirrel.Select(
+		"count(distinct session_id) as view_count").
+		From("viewership_events").
+		Where("user_id = ?", spec.Filter.UserID).
+		Where("server_timestamp > (toUnixTimestamp(now() - 30)) * 1000").
+		Limit(maxClickhouseResultRows)
+}
+
+// timeRangeEventsQuery uses an optimized materialized view "per minute" to favor query time against the latency.
+func timeRangeEventsQuery(spec QuerySpec) squirrel.SelectBuilder {
+	query := squirrel.Select(
+		"timestamp_ts",
+		"count(distinct session_id) as view_count",
+		"sum(buffer_ms) / sum(playtime_ms) as buffer_ratio",
+		"sum(if(errors > 0, 1, 0)) / count(*) as error_rate").
+		From("viewership_sessions_by_minute").
+		Where("user_id = ?", spec.Filter.UserID).
+		GroupBy("timestamp_ts").
+		OrderBy("timestamp_ts desc").
+		Limit(maxClickhouseResultRows)
+
+	if spec.From != nil {
+		query = query.Where("timestamp_ts >= ?", spec.From.UnixMilli()/1000)
+	}
+	if spec.To != nil {
+		query = query.Where("timestamp_ts < ?", spec.To.UnixMilli()/1000)
+	}
+	return query
+}
+
+func replaceNaNBufferRatio(rows []RealtimeViewershipRow) []RealtimeViewershipRow {
+	var res []RealtimeViewershipRow
+	for _, r := range rows {
+		if math.IsNaN(r.BufferRatio) {
+			r.BufferRatio = 1.0
```

Comment:
why do you use 1.0? 

is there a way to return missing values? The only way this should happen is if buffer and playtime are both equal to 0, i.e. `0 / 0` in which case a None value would be the most appropriate

Reply:
I think we have different options:
1. Return `null`
2. Return `0.0` (default float value)
3. Return `1.0` (since it's the closest to, the buffer ratio was super high)
4. Do not return this field at all

My intuition is as well to have "no problems here", which is `0.0`. Let's use it for now. But we can keep on discussing if that causes some confusion.

Reply:
Updated in 206ac95872815f3ecb2d76bc3e1a151e38f7e08b